core/shared/src/main/scala/org/http4s/Service.scala

```diff
@@ -0,0 +1,88 @@
+package org.http4s
+
+import cats.Applicative
+import cats.Monad
+import cats.MonoidK
+import cats.StackSafeMonad
+import cats.effect.kernel.Resource
+import cats.mtl.Local
+
+import Service._
+
+sealed abstract class Service[F[_], -A, B] {
+  def run: A => Resource[F, Option[B]]
+
+  def map[C](f: B => C): Service[F, A, C] =
+    apply((a: A) => run(a).map(_.map(f)))
+
+  def flatMap[AA <: A, C](f: B => Service[F, AA, C]) =
+    apply((a: AA) =>
+      run(a).flatMap(_ match {
+        case Some(b) => f(b).run(a)
+        case None => Resource.pure[F, Option[C]](None)
+      })
+    )
+
+  def orElse[AA <: A](default: => Service[F, AA, B]): Service[F, AA, B] =
+    apply(a =>
+      run(a).flatMap {
+        case some: Some[_] => Resource.pure(some)
+        case None => default.run(a)
+      }
+    )
+
+  def local[C](f: C => A): Service[F, C, B] =
+    apply(c => run(f(c)))
+}
+
+object Service extends ServiceInstances {
+  private[Service] final case class Run[F[_], -A, B](run: A => Resource[F, Option[B]])
```

Comment:
We can optimize this with a PartialFunction-like encoding.

Reply:
How would that work? `PartialFunction` has:
```
def isDefinedAt(x: A): Boolean
```
which seems problematic in an effectual setting. Even if it's `F[Boolean]`.

Reply:
See [01275e4](https://github.com/http4s/http4s/pull/6769/commits/01275e4896c8dd5b9b16c30d6c99f4615aa798e1).  I only compiled it, with a lot of shady casts, so it might not work at all.

Reply:
Fixed in [e8e380f](https://github.com/http4s/http4s/pull/6769/commits/e8e380fdc46fdd750db66a9b7b68a2f001854763)